README.md
# fetch-site
Fetch website with all the resources/responses/requests to local files, using [puppeteer](https://github.com/GoogleChrome/puppeteer/).

[![npm](https://img.shields.io/npm/v/fetch-site.svg "Version")](https://www.npmjs.com/package/fetch-site)
[![Build Status](https://travis-ci.org/futurist/fetch-site.svg?branch=master)](https://travis-ci.org/futurist/fetch-site)


## Install

**CLI**

```sh
npm install -g fetch-site
```

**Module**
```sh
npm install --save fetch-site
```

## Usage

**CLI Usage**

```sh
$ fetch-site url [options]

Options
--version           Show version info
--help              Show help info
--no-headless, -h   Set {headless: false} for 'launch-option'
--dir, -d           Dir to save result to
--shot, -s          Filename to save a screenshot after page open
--user-agent, -u    Set userAgent, string
--viewport, -v      Set viewport, e.g. `'{width:1024, height: 768}'`
--timeout, -t       Set maximum navigation time in milliseconds
--cookies, -c       Set cookies
--wait-for, -w      Wait for milliseconds/selector/function/closed(true)
--index-file        Default name of index file, like index.html
--on-response       onResponse event, function(response) as string
--launch-option, -l Launch option passed into puppeteer, object as string
--open-option, -o   Open option to passed into page, object as string
--on-before-open    Before open page event, function(page) as string
--on-after-open     After open page event, function(page) as string
--on-finish         Finish fetch event, function(page) as string

Examples
$ fetch-site http://baidu.com -h -w -o '{waitUntil:"networkidle0"}' -u 'My-UA-String' -t 0
```

Above example will open the url with `{headless:false}`, wait until `networkidle0`, and wait for page close to exit, using UAString `"My-UA-String"`.

**Module Usage**

```js
const fetchSite = require('fetch-site')
fetchSite({
  url: 'http://www.baidu.com',
  // whether to save a screenshot
  shot: 'shot.png',
  dir: 'baidu.com',
  launchOption:{
    headless: false
  },
  openOption:{
    timeout: 100*1e3,
    waitUntil: 'networkidle0'
  },
  onResponse: async response=>{
    if(/his/.test(response.url)) return false
    // return false will skip the resource
  },
  onBeforeOpen: async page=>{
    page.setViewport({width: 1440, height: 1000})
  },
  onAfterOpen: async page=>{
    await new Promise(r=>setTimeout(r, 5000))
  },
  onFinish: async page=>console.log('ok')
})
```

Then check the folder `baidu.com` for results.



CHANGELOG.md
<a name="1.8.2"></a>
## [1.8.2](https://github.com/futurist/fetch-site/compare/1.8.1...1.8.2) (9/23/2019, 4:21:06 AM)

- cleanup options  [ac95d26](https://github.com/futurist/fetch-site/commit/ac95d26)
- add isMain and isNavigation  [d258c38](https://github.com/futurist/fetch-site/commit/d258c38)
- add executablePath  [9570db1](https://github.com/futurist/fetch-site/commit/9570db1)


<a name="1.8.1"></a>
## [1.8.1](https://github.com/futurist/fetch-site/compare/1.7.9...1.8.1) (2019-9-4 20:18:41)

- add close check  [33c57ac](https://github.com/futurist/fetch-site/commit/33c57ac)
- 1.8.0  [cdbf672](https://github.com/futurist/fetch-site/commit/cdbf672)


<a name="1.7.9"></a>
## [1.7.9](https://github.com/futurist/fetch-site/compare/1.7.8...1.7.9) (2019-8-27 20:06:50)

- fix bug of getUA  [8c0c1ae](https://github.com/futurist/fetch-site/commit/8c0c1ae)


<a name="1.7.8"></a>
## [1.7.8](https://github.com/futurist/fetch-site/compare/1.7.7...1.7.8) (2019-8-22 10:11:42)

- add ignoreHTTPSErrors  [851fbaa](https://github.com/futurist/fetch-site/commit/851fbaa)
- update ua  [4e017c7](https://github.com/futurist/fetch-site/commit/4e017c7)
- reponse.json add timestamp  [f6431ed](https://github.com/futurist/fetch-site/commit/f6431ed)
- fix: extensionDir misconfig  [83a9764](https://github.com/futurist/fetch-site/commit/83a9764)
- add inject of message response of pptrBinding  [a9f5513](https://github.com/futurist/fetch-site/commit/a9f5513)
- add extension support and extensionDir  [cdcfc78](https://github.com/futurist/fetch-site/commit/cdcfc78)
- add urlToPath  [cefb7c8](https://github.com/futurist/fetch-site/commit/cefb7c8)


<a name="1.7.7"></a>
## [1.7.7](https://github.com/futurist/fetch-site/compare/1.7.6...1.7.7) (2019-8-6 17:35:20)

- fix: invalid url throw  [076daf8](https://github.com/futurist/fetch-site/commit/076daf8)


<a name="1.7.6"></a>
## [1.7.6](https://github.com/futurist/fetch-site/compare/1.7.5...1.7.6) (2019-8-6 17:13:32)

- refactor: capture new page  [e28e8bb](https://github.com/futurist/fetch-site/commit/e28e8bb)
- add revisionHash for conflicts when ENOTDIR  [d5bfebd](https://github.com/futurist/fetch-site/commit/d5bfebd)


<a name="1.7.5"></a>
## [1.7.5](https://github.com/futurist/fetch-site/compare/1.7.4...1.7.5) (2019-8-5 23:41:30)

- add auto default useragent  [c07da82](https://github.com/futurist/fetch-site/commit/c07da82)


<a name="1.7.4"></a>
## [1.7.4](https://github.com/futurist/fetch-site/compare/1.7.3...1.7.4) (2019-8-3 23:44:41)

- revert cli timeout=0  [c5f3fa6](https://github.com/futurist/fetch-site/commit/c5f3fa6)
- fix: timeout=0 not set bug  [8226752](https://github.com/futurist/fetch-site/commit/8226752)


<a name="1.7.3"></a>
## [1.7.3](https://github.com/futurist/fetch-site/compare/1.7.2...1.7.3) (2019-8-3 23:35:12)

- cli h,w default to true  [0d127e4](https://github.com/futurist/fetch-site/commit/0d127e4)
- remove yarn.lock  [02feed4](https://github.com/futurist/fetch-site/commit/02feed4)
- update package.json  [4830cdc](https://github.com/futurist/fetch-site/commit/4830cdc)
- using snapshots test for response.json; fix onload test bug  [d75a783](https://github.com/futurist/fetch-site/commit/d75a783)
- add __tests__/__snapshots__/local.js.snap  [e2eed0c](https://github.com/futurist/fetch-site/commit/e2eed0c)
- update test  [130f3f4](https://github.com/futurist/fetch-site/commit/130f3f4)
- fix ci error  [32ee302](https://github.com/futurist/fetch-site/commit/32ee302)
- update ignore rule  [e1ac91b](https://github.com/futurist/fetch-site/commit/e1ac91b)




__tests__/baidu.js
const main = require('../')

test('should work without throw', async ()=>{
  // EX=External Site
  if(!process.env.ES) return
  await main({
    url: 'http://www.baidu.com',
    shot: 'shot.png',
    dir: 'baidu.com',
    launchOption:{
      // headless: false
    },
    openOption:{
      timeout: 100*1e3,
      waitUntil: 'networkidle0'
    },
    onResponse: response=>{
      if(/his/.test(response.url)) return false
    },
    onBeforeOpen: async page=>{
      page.setViewport({width: 1440, height: 1000})
    },
    onAfterOpen: async page=>{
      await new Promise(r=>setTimeout(r, 100))
    },
    onFinish: async page=>expect(page).toBeTruthy()
  })
}, 30*1e3)


__tests__/local.js
const main = require('../')
const {readFileSync, writeFileSync} = require('fs')
const {promisify} = require('util')
const dirCompare = require('dir-compare')
const {exec} = require('child_process')
const sortOn = require('sort-on')

const run = promisify(exec)
const timeout = 30 * 1e3
const fixtures = __dirname + '/fixtures'
const launchOption = {
  args: ['--no-sandbox'],
}
function folderIsDiff(a,b, {noblob=true}={}){
  // makeChangeForTest(a)
  // makeChangeForTest(b)
  let excludeFilter = 'response.json,shot.png'
  if(noblob) excludeFilter+=',blob'
  return dirCompare.compare(a,b,{
    compareContent: true,
    excludeFilter
    // TODO: check isEqual these json/png files
    // CI will fail for screenshot.png since font-different, etc.
  }).then(result=>{
    for(let obj of result.diffSet){
      if(obj.state!='equal') return obj
    }
  })
}

function responseIsSame(responseFilePath) {
  const unstableResHeaders = [
    'etag',
    'date',
    'last-modified',
    'connection',
    'keep-alive'
  ]
  const unstableReqHeaders = [
    'x-devtools-emulate-network-conditions-client-id',
    'user-agent'
  ]
  let res = require(responseFilePath)
    .sort((a,b)=>a.file > b.file)
    .filter(v=> /^http/i.test(v.url))

  res.forEach(v=> {
    unstableResHeaders.forEach(x=>delete v.headers[x])
    unstableReqHeaders.forEach(x=>v.request && delete v.request.headers[x])
  })
  expect(res).toMatchSnapshot()
}

function makeChangeForTest(folder){
  const file = folder+'/response.json'
  let newResponse = JSON.parse(readFileSync(file,'utf8'))
  newResponse = sortOn(newResponse, ['file', 'headers.date'])
  newResponse.map(v=>{
    v.headers['date'] = null
    v.headers['x-devtools-emulate-network-conditions-client-id'] = null
    v.request.headers['x-devtools-emulate-network-conditions-client-id'] = null
  })
  writeFileSync(file, JSON.stringify(newResponse), 'utf8')
}

function setupServer(done){
  const toxy = require('toxy')
  const poisons = toxy.poisons
  const rules = toxy.rules
  const express = require('express')
  const app = express()
  let count = 0
  app.get('/', (req,res)=>res.redirect('index.html'))
  app.get('/count', (req,res)=>res.end(++count+''))
  app.use(express.static(fixtures+'/www'))
  const local = app.listen(18181, done)
  
  /** Simulate bandwidth */
  var proxy = toxy()
  // proxy
  // .all('/*')
  // .forward('http://localhost:18080')
  // .outgoingPoison(poisons.bandwidth({ bps: 10 }))
  // proxy.listen(18181)

  return {
    proxy, local
  }
}
// var fixtures=__dirname + '/fixtures'; setupServer()

beforeEach(done=>{
  this.folder = fixtures+'/tmp/'+ +new Date
  this.server = setupServer(done)
})

afterEach( done=>{
  const {server, folder} = this
  if(server){
    server.proxy.close()
    server.local.close()
  }
  // exec(`rm -rf ${folder}`, done)
  done()
})


describe('local site test', ()=>{
  test('should throw', async ()=>{
    expect.assertions(3)
    try{
      await main()
    }catch(e){
      expect(e).toBeTruthy()
    }

    try{
      await main({
        dir: 'tmp/local',
      })
    }catch(e){
      expect(e).toBeTruthy()
    }

    await main({
      url: 'http://localhost:18181',
      launchOption
    })
    await run(`rm -rf localhost_18181`)
    expect(1).toBeTruthy()
  }, timeout)


  test('onload', async ()=>{
    await main({
      dir: this.folder,
      url: 'localhost:18181',
      launchOption,
      openOption: {
        waitUntil:'domcontentloaded'
      },
      // this not worked with localhost??
      // onAfterOpen: page=> page.removeListener('response', page.fetchSite.responseHook)

      onBeforeOpen: async page=>{
        await page.exposeFunction('loaded', function(){
          console.log('loaded')
          page.removeListener('response', page.fetchSite.responseHook)
        })
        await page.evaluateOnNewDocument(function(){
          // console.log(location.href)
          // this event will emit 4 times:
          // /index.html, about:blank, [domloaded], /b/, blob:
          if(location.href.indexOf('/index.html')<0) return
          document.addEventListener('DOMContentLoaded', e=>{
            window.loaded()
          })
        })
      }
    })

    expect(
      await folderIsDiff(this.folder, fixtures+'/local-onload')
    ).toBeFalsy()

    responseIsSame(this.folder+'/response.json')

  }, timeout)

  test('networkidle2', async ()=>{
    await main({
      dir: this.folder,
      url: 'localhost:18181',
      launchOption,
      openOption: {
        waitUntil:'networkidle2'
      }
    })

    expect(
      await folderIsDiff(this.folder, fixtures+'/local-networkidle2')
    ).toBeFalsy()

    responseIsSame(fixtures+'/local-networkidle2/response.json')

  }, timeout)

  test('event hooks', async ()=>{
    await main({
      dir: this.folder,
      url: 'localhost:18181',
      launchOption,
      shot: 'shot.png',
      openOption: {
        waitUntil:'networkidle2'
      },
      onResponse: res=>{
        if(res.url.indexOf('dot')>-1) return false
        if(/^blob/.test(res.url)){
          res.url = 'blob:blob-url'
        }
      },
      onBeforeOpen: async page=>{
        page.setViewport({width: 1440, height: 1000})
      },
      onAfterOpen: async page=>{
        await new Promise(r=>setTimeout(r, 100))
      },
      onFinish: async page=>expect(page).toBeTruthy()
    })

    expect(
      await folderIsDiff(this.folder, fixtures+'/local-events', {noblob:false})
    ).toBeFalsy()

    responseIsSame(fixtures+'/local-events/response.json')

  }, timeout)

})



__tests__/fixtures/local-onload/http/localhost%3A18181/js/a.js
function addCss (fileName) {
  var head = document.head
  var link = document.createElement('link')

  link.type = 'text/css'
  link.rel = 'stylesheet'
  link.href = fileName

  head.appendChild(link)
}

window.onload = e=> setTimeout(()=>{
  addCss('css/a.css')
  document.getElementById('h1').style.background = 'url(images/中文.gif)'
}, 100)




index.js
const {URL} = require('url')
const fs = require('fs')
const {join:joinPath, dirname, extname} = require('path').posix
const {promisify} = require('util')
const puppeteer = require('puppeteer')
const makeDir = require('make-dir')
const mime = require('mime-types')
const revisionHash = require('rev-hash');

const readFile = promisify(fs.readFile)
const writeFile = promisify(fs.writeFile)
const renameFile = promisify(fs.rename)
const {assign} = Object

async function main({
  dir,
  url,
  shot,
  userAgent,
  viewport,
  timeout,
  cookies,
  launchOption,
  extensionDir,
  executablePath, // = '/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome',
  userDataDir, // = '/Users/jamesyang/Library/Application\ Support/Google/Chrome',
  waitFor,
  onResponse,
  indexFile = 'index.html',
  openOption,
  onBeforeOpen,
  onAfterOpen,
  onFinish,
}={}) {
  // check options
  if(!url) throw 'url cannot be empty'
  // 1. puppeteer cannot run inside Docker, since the sandbox
  // launchOption = assign({args: ['--no-sandbox']}, launchOption)
  // 2. shot will add await so `load` event not accurate
  // shot = shot || 'screenshot.png'
  openOption = assign({waitUntil: 'networkidle2'}, openOption)
  url = ensureHTTP(url)
  if(!dir){
    dir = hostDir(new URL(url).host)
  }

	try{
		const baseDir = joinPath(__dirname, 'extensions')
		const extensionName = fs.readdirSync(baseDir).filter(v=>fs.lstatSync(joinPath(baseDir, v)).isDirectory())[0]
		if(extensionName){
			extensionDir = joinPath(baseDir, extensionName)
		}
	}catch(e){}
	const CRX_PATH = extensionDir
	// launch puppeteer
	launchOption = assign({
		defaultViewport: viewport,
		handleSIGINT: false,
		ignoreHTTPSErrors: true,
		args: [
			userAgent && `--user-agent=${userAgent}`,
			'--disable-infobars',
			'--no-sandbox',
			'--disable-setuid-sandbox',
			'--ignore-certifcate-errors',
			'--ignore-certifcate-errors-spki-list',
			// '--disable-web-security',
			'--enable-devtools-experiments',
			// '--auto-open-devtools-for-tabs',
			extensionDir && `--disable-extensions-except=${CRX_PATH}`,
			extensionDir && `--load-extension=${CRX_PATH}`
		].filter(Boolean)
	}, launchOption, executablePath && {executablePath}, userDataDir && {userDataDir})
  let browser = await puppeteer.launch(launchOption)
  await makeDir(dir)
	const page = (await browser.pages())[0] || await browser.newPage()
	const responseData = []

	const writeData = async ({page, url, data, getBuffer})=>{
    if((/^data:/i.test(url))
      || onResponse && await onResponse(data, page)===false
		) return

		// if(!page.isRoot) console.log(url)

    let body
    try{
      body = await getBuffer()
    }catch(e){
      // for 301/302 redirect, will throw no_body
      responseData.push(data)
      return
    }
    // reset url from onResponse
    if(url!==data.url) {
      data.realUrl = url
      url = data.url
    }
    const contentType = data.headers['content-type']
    const extension = mime.extension(contentType) // inferred from mime
		const charset = mime.charset(contentType)
		const pathArr = urlToPath(url, indexFile)
    if(!pathArr) return

    // push data
    const file = data.file = joinPath(...pathArr)
    responseData.push(data)

    // write to file
    const writeOption = {
      encoding: toNodeEncoding(charset)
    }
    let filePath = joinPath(dir, file)
    await ensureFolder(filePath, indexFile)
    try{
      let stat = fs.lstatSync(filePath)
      if (stat.isDirectory()) {
        filePath = joinPath(filePath, indexFile)
      }
      stat = fs.lstatSync(filePath)
      if(stat.isFile()) {
        if(body.equals(await readFile(filePath))){
          // it's same file, do nothing
          return
        }
        const suffix = '.'+(+new Date).toString(36)
        filePath += suffix
        data.file += suffix
      }
    }catch(e){
      // ENOENT -2 (not exists) is OK
		}
		try {
			await writeFile(filePath, body, writeOption)
		} catch(e) {
			/**
			 * Error: ENOTDIR: not a directory
			 * errno: -20, code: 'ENOTDIR', syscall: 'open',
			 */
			const name = revisionHash(body)
			const file = data.file = joinPath('conflicts', name)
			filePath = joinPath(dir, file)
			await ensureFolder(filePath, indexFile)
			await writeFile(filePath, body, writeOption)
		}
	}
  // response hook
  const responseHook = page => async response => {
    const req = response.request()
		let url = response.url()
    const request = {
      method: req.method(),
      headers: req.headers(),
			postData: req.postData(),
			isMain: req.frame() && !req.frame().parentFrame(),
      isNavigation: req.isNavigationRequest(),
      resourceType: req.resourceType(),
      // redirects: req.redirectChain().map(v=>v.url())
    }
    // const reqUrl = req.url()
    // if(reqUrl != url) request.url = reqUrl
    const data = {
      url,
      file: '',  // optimize v8 hidden class
      status: response.status(),
      ok: response.ok(),
      headers: response.headers(),
      request
		}
		await writeData({
			page,
			data,
			url,
			getBuffer: ()=>{
				return response.buffer()
			}
		})
	}
	const whenClose = page => new Promise(res=>{
    page.on('close', res)
    page.on('error', res)
	})
	const hookPage = page =>{
		page.fetchSite = {}
		const res = responseHook(page)
		page.on('response', res)
		whenClose(page).then(()=>{
			page.off('response', res)
		})
		const inject = function(){
			delete navigator.__proto__.webdriver
			window.addEventListener('message', async ({data, ports, target})=>{
				const port = ports && ports[0]
				target = port || target
				if (data) {
					const {type} = data
					if(type === 'pptrBinding') {
						const payload = await window[data.method](data.params)
						if(target instanceof Worker || target instanceof MessagePort) {
							target.postMessage(payload)
						} else if(target) {
							target.postMessage(payload, '*')
						}
					}
				}
				port && port.close()
			})
		}
		page.evaluate(inject)
		page.evaluateOnNewDocument(inject)
    page.exposeFunction('__pptr__fileUrlExist', async e => {
      const {url} = e
      const pathArray = urlToPath(url, indexFile)
      if(!pathArray) {
        return
      }
      const filePath = joinPath(dir, ...pathArray)
      try{
        let stat = fs.lstatSync(filePath)
        return stat
      }catch(e){
        // console.log(e)
      }
    })
	}
	hookPage(page)
	page.isRoot = true

	browser.on('targetcreated', async target => {
		if (target.type() === 'page') {
      if (/^https?:\/\//i.test(target.url())) {
				console.log('new tab created:', target.url())
				let page
        try {
          page = await target.page()
					hookPage(page)
					page.setCacheEnabled(false)
					// 'response' only applied to resource
					const isLoaded = await page.evaluate(()=>{
						return document.readyState != 'loading'
					})
					const run = async()=>{
						const url = page.url()
						const request = {
							method: 'GET',
							headers: {},
							postData: {},
							resourceType: 'document',
							// redirects: req.redirectChain().map(v=>v.url())
						}
						// const reqUrl = req.url()
						// if(reqUrl != url) request.url = reqUrl
						const data = {
							url,
							file: '',  // optimize v8 hidden class
							status: 200,
							ok: true,
							headers: {},
							request
						}
						await writeData({
							page,
							data,
							url,
							getBuffer: ()=>{
								return page.content()
							}
						})
					}
					if(isLoaded){
						run()
					} else {
						page.on('domcontentloaded', run)
					}
        } catch (e) {
          console.log(e.message)
          if (/Target closed/i.test(e.message)) {
            return
          }
          if (/No target with given id found/i.test(e.message)) {
            return
          }
        }
			}
		}
	})

  // goto url
  if(userAgent) page.setUserAgent(userAgent)
  if(viewport) page.setViewport(viewport)
  if(timeout != null) page.setDefaultNavigationTimeout(timeout)
  if(cookies) page.setCookie(...cookies)
	onBeforeOpen && await onBeforeOpen(page)
	try{
		await page.goto(url, openOption)
	}catch(e){
		console.log(e.message)
	}
  // wait finish
  const pending = []
  onAfterOpen && pending.push(onAfterOpen(page))
  if(waitFor){
    pending.push(waitFor===true ? whenClose(page) : page.waitFor(waitFor))
	}

	const whenEnd = async ()=>{
		const filename = `response-${Date.now()}.json`
		await writeFile(joinPath(dir, filename), JSON.stringify(responseData), 'utf8')
		console.log(`wrote ${filename}`)
		// screen shot
		if(shot){
			if(shot===true) shot = 'screenshot.png'
			const opt = typeof shot==='object'
			? shot
			: {
				path: joinPath(dir, shot),
				fullPage: true
			}
			try{
				// ensure page not closed
				await page.screenshot(opt)
			}catch(e){}
		}

		// finish work
		let alive = browser && !(browser).disconnected
		if(alive){
			try {
				await browser.close()
			} catch(e){}
		}
		onFinish && await onFinish(page)

		return {
			dir,
			url,
			data: responseData
		}
	}

	let closePromiseRes
	const closePromise = new Promise((res)=>closePromiseRes = res)
	async function checkClosed () {
		let alive = browser && !(browser).disconnected
		if (alive) {
			// browser.disconnect()
			await browser.close()
			browser = null
			console.log('browser closed')
		}
		closePromiseRes()
	}

	browser.on('disconnected', e => {
    browser.removeAllListeners()
    browser.disconnected = true
  })

	process.on('beforeExit', () => checkClosed())
  process.on('SIGTERM', () => checkClosed())
  process.on('SIGINT', () => checkClosed())

  await Promise.race([
    Promise.all(pending),
		whenClose(page),
		closePromise
  ])

	return whenEnd()
}

process.on('unhandledRejection', console.log)

async function ensureFolder(filePath, indexFile) {
  try{
    await makeDir(dirname(filePath))
  } catch(e){
    let {code, path} = e
    if(!path) return
    // e.g. exist file: xx/abc
    // folder to create: xx/abc/x/y
    // will throw errno:-20
    if(code==='ENOTDIR') {
      await ensureFolder(path, indexFile)
    }
    // folder to create: xx/abc/x
    // will throw errno:-17
    if(code==='EEXIST') {
      const tempPath = path +
        Math.random().toString(36).slice(1)
      await renameFile(path, tempPath)
      await makeDir(path)
      await renameFile(tempPath, joinPath(path, indexFile))
    }
    await ensureFolder(filePath, indexFile)
  }
}

function hostDir(host, replacement = '_') {
  return host.replace(':', replacement)
}

// prevent multiple escape % as %25
function safePath (url) {
  return encodeURIComponent(url).replace(/%25/g, '%')
}

function ensureHTTP(url){
  return !/^https?:\/\//i.test(url)
    ? 'http://'+url
    : '' + url
}

function ensureIndex(filePath, indexFile){
  if(filePath.endsWith('/')) filePath += indexFile
  return filePath
}

function urlToPath(url, indexFile) {
  let urlObj = {}
		try{
			urlObj = new URL(url)
		} catch(e){
			// TypeError [ERR_INVALID_URL]: Invalid URL: ":"
			return
		}
    const {pathname, protocol, host} = urlObj
    const pathArr = [
      hostDir(protocol, ''),
      hostDir(host),
    ].concat(
      /^blob/.test(protocol)  // url as a whole
        ? safePath(pathname)
        : ensureIndex(pathname, indexFile).split('/').map(safePath)
    )
  return pathArr
}

function toNodeEncoding(enc){
  return enc && typeof enc==='string'
    ? enc.toLowerCase().replace('-', '')
    : 'utf8'
}

module.exports = main


cli.js
#!/usr/bin/env node

const {URL} = require('url')
const evalExpression = require('eval-expression')
const main = require('./')
const pkg = require('./package.json')
const {getUA} = require('./ua')

const cli = require('meow')(`
Usage
$ ${pkg.name} url [options]

Options
--version            Show version info
--help               Show help info
--no-headless, -h    Set {headless: false} for 'launch-option', default true
--dir, -d            Dir to save result to
--shot, -s           Filename to save a screenshot after page open
--user-agent, -u     Set userAgent, string
--executable-path    Set executablePath for browser
--user-data-dir      Set userDataDir for browser
--viewport, -v       Set viewport, , e.g. '{width:1024, height: 768}'
--timeout, -t        Set maximum navigation time in milliseconds
--cookies, -c        Set cookies
--wait-for, -w       Wait for milliseconds/selector/function/closed, default true
--index-file         Default name of index file, like index.html
--extension-dir      Extension dir
--on-response        onResponse event, function(response) as string
--launch-option, -l  Launch option passed into puppeteer, object as string
--open-option, -o    Open option to passed into page, object as string
--on-before-open     Before open page event, function(page) as string
--on-after-open      After open page event, function(page) as string
--on-finish          Finish fetch event, function(page) as string

Examples
$ ${pkg.name} http://baidu.com -o '{waitUntil:"networkidle0"}'
`, {
  flags:{
    dir: {alias: 'd'},
    shot: {alias: 's'},
    'wait-for': {alias: 'w', default: true},
    'user-agent': {alias: 'u'},
    'viewport': {alias: 'v'},
    'timeout': {alias: 't'},
    'cookies': {alias: 'c'},
    'no-headless': {type: 'boolean', alias: 'h', default: true},
    'open-option': {alias: 'o'},
    'extension-dir': {},
    'launch-option': {alias: 'l'},
  }
})

const {flags, input} = cli
const [url] = input||[]

if(!url) cli.showHelp()
else flags.url = url

const ua = getUA(flags.u)
flags.userAgent = ua ? ua.useragent : flags.u

;[
  'launchOption',
  'openOption',
  'onBeforeOpen',
  'onAfterOpen',
  'onFinish',
  'onResponse',
  'waitFor',
  'cookies',
].forEach(v=>flags[v] = evalExpression(flags[v]))

flags.launchOption = Object.assign({
	headless: !flags.noHeadless,
	defaultViewport: null
}, flags.launchOption)

;(async ()=>{
  const {dir, url, data} = await main(flags) || {}
  console.log(`\nSuccess: site ${url} saved into ${dir}`)
})()

